File: 14_back/13_8.c
#include <stdio.h>
#include <stdlib.h>

int stat[21][21] = {0, };
int	flag[21] = {0, };
int	team = 0;
int	min = 10000;

void	startlink(int n, int index, int combi)
{
	int	start = 0;
	int link = 0;

	if (team == index / 2)
	{
		for (int i = 1; i <= index; i++)
		{
			for (int j = 1; j <= index; j++)
			{
				if (flag[i] == 0 && flag[j] == 0)
					start += stat[i][j];
				else if (flag[i] == 1 && flag[j] == 1)
					link += stat[i][j];
			}
		}
	
		if (abs(start - link) < min)
			min = abs(start - link);
		
		return ;
	}

	for (int i = 1; i <= index; i++)
	{
		if (i > combi && flag[i] == 0)
		{
			flag[i] = 1;
			team++;

			startlink(n + 1, index, i);

			if (min == 0)
				break ;
			
			flag[i] = 0;
			team--;
		}
	}
}

int	main(void)
{
	int	n;

	scanf("%d", &n);

	for (int i = 1; i <= n; i++)
		for (int j = 1; j <= n; j++)
			scanf("%d", &stat[i][j]);
	
	startlink(1, n, 0);

	printf("%d\n", min);
}

File: 14_back/13_8.py
from itertools import combinations

num = int(input())
stat = [list(map(int, input().split())) for _ in range(num)]
team = []

for i in range(0, num, 1):
	team.append(i)

startlink = list(combinations(team, num // 2))
start = 0
link = 0
mini = 10000

for i in range(0, len(startlink) // 2, 1):
	start = 0
	link = 0
	for j in range(0, (num // 2) - 1, 1):
		for k in range(j + 1, num // 2, 1):
			row = startlink[i][j]
			col = startlink[i][k]
			l_row = startlink[len(startlink) - i - 1][j]
			l_col = startlink[len(startlink) - i - 1][k]
			start += stat[row][col] + stat[col][row]
			link += stat[l_row][l_col] + stat[l_col][l_row]
	if (abs(start - link) < mini):
		mini = abs(start - link)
	if (mini == 0):
		break

print(mini)


